client-node-tests/src/integration.test.ts

```diff
@@ -1515,6 +1515,43 @@ suite('Client integration', () => {
 		middleware.sendNotification = undefined;
 		assert.strictEqual(middlewareCallCount, 2);
 	});
+
+	test('applyEdit middleware', async () => {
+		const middlewareEvents: Array<lsclient.ApplyWorkspaceEditParams> = [];
+		let currentProgressResolver: (value: unknown) => void | undefined;
+
+		// Set up middleware that calls the current resolve function when it gets its 'end' progress event.
+		middleware.workspace = middleware.workspace || {};
+		middleware.workspace.handleApplyEdit = async (params, next) => {
+			middlewareEvents.push(params);
+			setImmediate(currentProgressResolver);
+			/**
+			 * TODO signatures need to be updated also for showDocument, handleRegisterCapability, handleUnregisterCapability
+			 * The use of `HandlerSignature` makes it that you have to pass a dummy cancellation token and handle `ResponseError`
+			 */
+			const resultOrError = await next(params, tokenSource.token);
+			if(resultOrError instanceof lsclient.ResponseError) {
+				return Promise.reject(resultOrError.message);
+			}
+			return resultOrError;
```

Comment:
It seems using `HandlerSignature` is not ideal for middlewares. We essentially have to pass a null `CancellationToken`  and handle `ResponseError`. Requests sent from the server to the client like `applyEdit`, `showDocument`, `handleRegisterCapability`, `handleUnregisterCapability` now have to do this workaround...

Am I using it wrong or is this expected?

Reply:
The token we have to provide since this is how things are speced right now. There could be servers that provide cancellation for message send from the server.

For the return value I think we should simply declare the middleware as returning a `HandlerResult`. The a middle ware implementation doesn't need to handle the ResponseError.

Let me know if this works

Reply:
Is it possible to make the token optional then? Requiring the token makes the code a bit awkward.

Just returned a `HandlerResult` and updated the middleware handling such that it returns a rejected promise. Should I update the middlewares of `showDocument`, `handleRegisterCapability`, `handleUnregisterCapability` such that they return a `HandlerResult` too (+ simple tests for sanity)?

Reply:
I decided to leave the middle ware as is for now since technically this might be breaking and I am not sure if it is worth the effort right now.

IMO the token is only a problem when writing the tests. In a real world scenario the token should already be provided.

Reply:
That's fair, but I tried implementing the middleware `handleRegisterCapability` in production and there is no way to propagate the cancellation token so I have to create a throwaway one.

I would expect the middleware to be able to follow this format: `handleX: (...args, next) => return next(...args)`.

For example, the `didOpen`:

```javascript
didOpen: async (data, next) => {
    return next(data)
}
```